File: qeli-shared/QeliShared/TransportAuto.cs
namespace Qeli.Shared;

/// <summary>Auto transport ranking — mirrors Rust <c>config::auto_transport</c>.</summary>
public static class TransportAuto
{
    public static readonly string[] DefaultOrder =
    {
        "reality-tls", "fake-tls", "obfs", "plain", "quic", "udp-quic", "udp-obfs", "udp"
    };

    /// <summary>Short default picker list (main-screen combo). Full catalog: <see cref="SniCatalog.Builtin"/>.</summary>
    public static readonly string[] SniPresets =
    {
        "www.cloudflare.com",
        "www.google.com",
        "www.microsoft.com",
        "www.apple.com",
        "www.amazon.com",
        "www.speedtest.net",
        "github.com",
        "cdn.jsdelivr.net",
        "www.bing.com",
        "login.live.com",
        "www.nvidia.com",
        "www.wikipedia.org",
    };

    public static string TransportLabel(string protocol, string mode, bool quic)
    {
        var proto = (protocol ?? "").Trim().ToLowerInvariant();
        var m = (mode ?? "").Trim().ToLowerInvariant();
        if (proto == "udp")
        {
            if (quic || m == "udp-quic") return "quic";
            if (m is "obfs" or "udp-obfs") return "udp-obfs";
            return "udp";
        }
        return m switch
        {
            "reality-tls" or "reality" => "reality-tls",
            "obfs" => "obfs",
            "plain" => "plain",
            "fake-tls" or "tls" or "" => "fake-tls",
            _ => m,
        };
    }

    public static int PreferenceRank(string label, IReadOnlyList<string>? order = null)
    {
        order ??= DefaultOrder;
        for (var i = 0; i < order.Count; i++)
            if (string.Equals(order[i], label, StringComparison.OrdinalIgnoreCase))
                return i;
        return 900 + label.Length;
    }

    public readonly record struct Ranked(int Index, long? RttMs, int Preference);

    public static List<Ranked> Rank(
        IReadOnlyList<string> labels,
        IReadOnlyList<bool> ok,
        IReadOnlyList<long?> rttMs,
        IReadOnlyList<string>? order = null)
    {
        order ??= DefaultOrder;
        var picks = new List<Ranked>();
        for (var i = 0; i < labels.Count; i++)
        {
            if (!ok[i]) continue;
            picks.Add(new Ranked(i, rttMs[i], PreferenceRank(labels[i], order)));
        }
        picks.Sort((a, b) =>
        {
            var ba = (a.RttMs ?? long.MaxValue) / 40;
            var bb = (b.RttMs ?? long.MaxValue) / 40;
            var c = ba.CompareTo(bb);
            if (c != 0) return c;
            c = a.Preference.CompareTo(b.Preference);
            if (c != 0) return c;
            c = (a.RttMs ?? long.MaxValue).CompareTo(b.RttMs ?? long.MaxValue);
            return c != 0 ? c : a.Index.CompareTo(b.Index);
        });
        return picks;
    }

    public static string ApplySniToIni(string ini, string sni)
    {
        var host = (sni ?? "").Trim();
        if (host.Length == 0) throw new ArgumentException("sni must not be empty");
        var lines = ini.Replace("\r\n", "\n").Replace('\r', '\n').Split('\n').ToList();
        var replaced = false;
        for (var i = 0; i < lines.Count; i++)
        {
            var t = lines[i].Trim();
            if (t.StartsWith("sni", StringComparison.OrdinalIgnoreCase) && t.Contains('='))
            {
                lines[i] = "sni = " + host;
                replaced = true;
                break;
            }
        }
        if (!replaced)
        {
            var insertAt = lines.FindIndex(l => l.Trim().Equals("[qeli]", StringComparison.OrdinalIgnoreCase));
            insertAt = insertAt >= 0 ? insertAt + 1 : lines.Count;
            lines.Insert(insertAt, "sni = " + host);
        }
        return string.Join("\n", lines);
    }
}

File: qeli-android/app/src/main/kotlin/com/qeli/TransportAuto.kt
package com.qeli

/**
 * Auto transport ranking — mirrors Rust `config::auto_transport`.
 * Preference: reality-tls → fake-tls → obfs → plain → quic.
 */
object TransportAuto {
    val DEFAULT_ORDER = listOf(
        "reality-tls", "fake-tls", "obfs", "plain", "quic", "udp-quic", "udp-obfs", "udp"
    )

    val SNI_PRESETS = listOf(
        "www.cloudflare.com",
        "www.google.com",
        "www.microsoft.com",
        "www.apple.com",
        "www.amazon.com",
        "www.speedtest.net",
        "github.com",
        "cdn.jsdelivr.net",
        "www.bing.com",
        "login.live.com",
        "www.nvidia.com",
        "www.wikipedia.org",
    )

    fun transportLabel(protocol: String, mode: String, quic: Boolean): String {
        val proto = protocol.trim().lowercase()
        val m = mode.trim().lowercase()
        if (proto == "udp") {
            if (quic || m == "udp-quic") return "quic"
            if (m == "obfs" || m == "udp-obfs") return "udp-obfs"
            return "udp"
        }
        return when (m) {
            "reality-tls", "reality" -> "reality-tls"
            "obfs" -> "obfs"
            "plain" -> "plain"
            "fake-tls", "tls", "" -> "fake-tls"
            else -> m
        }
    }

    fun preferenceRank(label: String, order: List<String> = DEFAULT_ORDER): Int {
        val idx = order.indexOfFirst { it.equals(label, ignoreCase = true) }
        return if (idx >= 0) idx else 900 + label.length
    }

    data class Ranked(val index: Int, val rttMs: Long?, val preference: Int)

    fun rank(
        labels: List<String>,
        ok: List<Boolean>,
        rttMs: List<Long?>,
        order: List<String> = DEFAULT_ORDER,
    ): List<Ranked> {
        require(labels.size == ok.size && ok.size == rttMs.size)
        val picks = ArrayList<Ranked>()
        for (i in labels.indices) {
            if (!ok[i]) continue
            picks += Ranked(i, rttMs[i], preferenceRank(labels[i], order))
        }
        picks.sortWith(
            compareBy<Ranked> { (it.rttMs ?: Long.MAX_VALUE) / 40 }
                .thenBy { it.preference }
                .thenBy { it.rttMs ?: Long.MAX_VALUE }
                .thenBy { it.index }
        )
        return picks
    }

    fun applySniToIni(ini: String, sni: String): String {
        val host = sni.trim()
        require(host.isNotEmpty()) { "sni must not be empty" }
        val lines = ini.replace("\r\n", "\n").replace('\r', '\n').lines().toMutableList()
        var replaced = false
        for (i in lines.indices) {
            val t = lines[i].trim()
            if (t.startsWith("sni", ignoreCase = true) && t.contains('=')) {
                lines[i] = "sni = $host"
                replaced = true
                break
            }
        }
        if (!replaced) {
            val insertAt = lines.indexOfFirst { it.trim().equals("[qeli]", ignoreCase = true) }
                .let { if (it >= 0) it + 1 else lines.size }
            lines.add(insertAt, "sni = $host")
        }
        return lines.joinToString("\n")
    }
}
